### PR TITLE
Fix PostgreSQL Cidr#change?: NoMethodError: undefined method `prefix' for nil

### PR DESCRIPTION
**Pull request by taketo1113**
### Motivation / Background

This Pull Request has been created because fixes an error.

Steps to reproduce are below comments.
https://github.com/rails/rails/pull/51633#issuecomment-2098218748

### Detail

PostgreSQL Cidr#change? raise an error of NoMethodError: undefined method `prefix' for nil, when creating a record with an empty value of inet/cidr column.

This Pull Request changes fixes the error and adds a test case.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [ ] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/test/cases/adapters/postgresql/cidr_test.rb [hunk 1]
```diff
@@ -21,6 +21,15 @@ class CidrTest < ActiveRecord::PostgreSQLTestCase
...
           assert_equal "foo", type.serialize("foo")
         end
+
+        test "changed? with nil values" do
+          type = OID::Cidr.new
+
+          assert_equal false, type.changed?(nil, nil, "")
+          assert_equal true, type.changed?("192.168.0.0/24", nil, "")
+          assert_equal true, type.changed?(nil, "192.168.0.0/24", "")
+          assert_equal true, type.changed?("192.168.0.0/24", "192.168.0.0/25", "")
```
Comment (carlosantoniodasilva) [inline activerecord/test/cases/adapters/postgresql/cidr_test.rb:31]: We probably don't need to verify equality with `false/true` here, can we just use `assert` and `assert_not` on these, to simplify?

Comment (taketo1113) [reply activerecord/test/cases/adapters/postgresql/cidr_test.rb]: @carlosantoniodasilva Thanks for comments.
I have fixed it.

**Comment by rafal-brize** [comment]
hey folks i believe this might have introduced some regression. observed on bump from 

I am seeing an error `NoMethodError: undefined method `prefix' for nil` on Rails 7.1.4
on the code that runs regular  `resource.update!(network_address: nil)`
I am having problems recreating it.

here is the stack

```
 {
          "method": "update!",
          "lineno": 904,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/persistence.rb"
        },
        {
          "method": "with_transaction_returning_status",
          "lineno": 361,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/transactions.rb"
        },
        {
          "method": "transaction",
          "lineno": 342,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/connection_adapters/abstract/database_statements.rb"
        },
        {
          "method": "block in with_transaction_returning_status",
          "lineno": 365,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/transactions.rb"
        },
        {
          "method": "block in update!",
          "lineno": 906,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/persistence.rb"
        },
        {
          "method": "save!",
          "lineno": 56,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/suppressor.rb"
        },
        {
          "method": "save!",
          "lineno": 313,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/transactions.rb"
        },
        {
          "method": "with_transaction_returning_status",
          "lineno": 361,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/transactions.rb"
        },
        {
          "method": "transaction",
          "lineno": 342,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/connection_adapters/abstract/database_statements.rb"
        },
        {
          "method": "block in with_transaction_returning_status",
          "lineno": 365,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/transactions.rb"
        },
        {
          "method": "block in save!",
          "lineno": 313,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/transactions.rb"
        },
        {
          "method": "save!",
          "lineno": 55,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/validations.rb"
        },
        {
          "method": "save!",
          "lineno": 751,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/persistence.rb"
        },
        {
          "method": "create_or_update",
          "lineno": 127,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/timestamp.rb"
        },
        {
          "method": "create_or_update",
          "lineno": 441,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/callbacks.rb"
        },
        {
          "method": "_run_save_callbacks",
          "lineno": 952,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activesupport-7.1.4/lib/active_support/callbacks.rb"
        },
        {
          "method": "run_callbacks",
          "lineno": 141,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activesupport-7.1.4/lib/active_support/callbacks.rb"
        },
        {
          "method": "block in run_callbacks",
          "lineno": 130,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activesupport-7.1.4/lib/active_support/callbacks.rb"
        },
        {
          "method": "around_save_collection_association",
          "lineno": 375,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/autosave_association.rb"
        },
        {
          "method": "block in run_callbacks",
          "lineno": 121,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activesupport-7.1.4/lib/active_support/callbacks.rb"
        },
        {
          "method": "block in create_or_update",
          "lineno": 441,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/callbacks.rb"
        },
        {
          "method": "create_or_update",
          "lineno": 1221,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/persistence.rb"
        },
        {
          "method": "_update_record",
          "lineno": 122,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/timestamp.rb"
        },
        {
          "method": "_update_record",
          "lineno": 449,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/callbacks.rb"
        },
        {
          "method": "_run_update_callbacks",
          "lineno": 952,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activesupport-7.1.4/lib/active_support/callbacks.rb"
        },
        {
          "method": "run_callbacks",
          "lineno": 110,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activesupport-7.1.4/lib/active_support/callbacks.rb"
        },
        {
          "method": "block in _update_record",
          "lineno": 449,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/callbacks.rb"
        },
        {
          "method": "record_update_timestamps",
          "lineno": 140,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/timestamp.rb"
        },
        {
          "method": "block (2 levels) in _update_record",
          "lineno": 449,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/callbacks.rb"
        },
        {
          "method": "_update_record",
          "lineno": 233,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/attribute_methods/dirty.rb"
        },
        {
          "method": "attribute_names_for_partial_updates",
          "lineno": 246,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/attribute_methods/dirty.rb"
        },
        {
          "method": "changed_attribute_names_to_save",
          "lineno": 182,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/attribute_methods/dirty.rb"
        },
        {
          "method": "changed_attribute_names",
          "lineno": 15,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activemodel-7.1.4/lib/active_model/attribute_mutation_tracker.rb"
        },
        {
          "method": "select",
          "lineno": 15,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activemodel-7.1.4/lib/active_model/attribute_mutation_tracker.rb"
        },
        {
          "method": "block in changed_attribute_names",
          "lineno": 15,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activemodel-7.1.4/lib/active_model/attribute_mutation_tracker.rb"
        },
        {
          "method": "changed?",
          "lineno": 45,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activemodel-7.1.4/lib/active_model/attribute_mutation_tracker.rb"
        },
        {
          "method": "attribute_changed?",
          "lineno": 79,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activemodel-7.1.4/lib/active_model/attribute_mutation_tracker.rb"
        },
        {
          "method": "changed?",
          "lineno": 67,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activemodel-7.1.4/lib/active_model/attribute.rb"
        },
        {
          "method": "changed_from_assignment?",
          "lineno": 162,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activemodel-7.1.4/lib/active_model/attribute.rb"
        },
        {
          "method": "changed?",
          "lineno": 34,
          "filename": "/app/vendor/bundle/ruby/3.3.0/gems/activerecord-7.1.4/lib/active_record/connection_adapters/postgresql/oid/cidr.rb"
        }
      ]
```

seams like `new_value` must be `nil` as `old_value` is already guarded against `nil`

@carlosantoniodasilva, @taketo1113 wondering if you have experienced that?

**Comment by rafaelfranca** [comment]
This PR has a test for that case https://github.com/rails/rails/commit/362124753d35fbb62354920f0bb22c2b1337cedf#diff-d4c368a29f44862fe2a0736d8aae9fe1938ecf61fb72f62957d980f6aa724184R29

**Comment by rafal-brize** [comment]
thanks @rafaelfranca, @taketo1113 that is indeed true, but please notice that those tests are done with strings, not instances of `IPAddr`

I have managed to reproduce the issue, took a while.

So

```
type = ActiveRecord::ConnectionAdapters::PostgreSQLAdapter::OID::Cidr.new
type.changed?(IPAddr.new("0.0.0.0/8"), nil, "")
```
raises 
`NoMethodError: undefined method 'prefix' for nil`

as well as end-2-end with AR

```
>> eu.update!(network_address: "0.0.0.0/8")
=> true
>> eu.update!(network_address: nil)
NoMethodError: undefined method `prefix' for nil
```

the culprit seems to be a hidden within IPAddr

https://github.com/ruby/ipaddr/blob/master/lib/ipaddr.rb#L155-L161

please take a look at that!

```
IPAddr.new("0.0.0.0") == nil 
=> true
```

I guess changing implementation of `IPAddr` is out of the question for backwards compatibility.
How do you want to handle that? another explicit nil check maybe?

**Comment by rafaelfranca** [comment]
That seems to be a bug in IPAddr. @taketo1113 already opened a PR to fix it. Let's watch that one. If they don't want to change the behavior we can change in Rails.

**Comment by taketo1113** [comment]
I reported Issue/PR of `IPAddr.new("0.0.0.0") == nil` in ipaddr gem.

@rafaelfranca I think it should be fix the behavior in Rails.
Even if fixed the behavior in IPAddr, it will take some time before it is available in Rails because IPAddr is the default gem.

**Comment by taketo1113** [comment]
@rafaelfranca I tried creating a PR.
- #52721

**Pull request merged**